server,sqlliveness: rework shutting down SQL pods on sqlliveness problems

andreimatei opened the pull request:
This patch changes how stopping the stopper and, in turn, terminating
the process happens for tenant SQL pods when the sqlliveness session is
expired. Before this patch, the liveness code was calling the
Stopper.Stop() directly. This patch replaces that with bubbling up an
event up to the cli layer, which is where the logic around terminating
the process lives. That guy has  more complex shutdown logic - it has
the opportunity to perform draining if it wants (or at least it will
in https://github.com/cockroachdb/cockroach/pull/90143), and it also prints
a message with the cause of the exit.

An alternative considered was to extend the Stopper.Stop() interface
somehow with a termination message, and continue to use the stopper as
the mechanism used by arbitrary modules to signal the desire to
terminate the server. I've decided against it, though, in favor of
introducing another mechanism for signaling requests to terminate up to
the cli layer, on the argument that the stopper is passed around so
widely and used in a lot of non-production libraries, such that it's
hard to trace and tell what production modules call Stop() on it and
under which circumstances. Also, stopper.Stop() cannot be deadlocks if
called from inside a stopper task (and our code generally runs inside
stopper tasks), so we'd also need to add a new async stop interface to
the stopper.
Thus, the goal of this patch is to keep the number of callers to
Stopper.Stop() small - right now there are two of them: one in the cli,
and one in the drainServer. In a followup patch, I might remove the
drainServer call, and then we'd be left with exactly one caller to
Stopper.Stop().

Fixes https://github.com/cockroachdb/cockroach/issues/85540

Release note: None
Epic: None

andreimatei committed:
server,sqlliveness: rework shutting down SQL pods on sqlliveness problems

This patch changes how stopping the stopper and, in turn, terminating
the process happens for tenant SQL pods when the sqlliveness session is
expired. Before this patch, the liveness code was calling the
Stopper.Stop() directly. This patch replaces that with bubbling up an
event up to the cli layer, which is where the logic around terminating
the process lives. That guy has  more complex shutdown logic - it has
the opportunity to perform draining if it wants (or at least it will
in #90143), and it also prints a message with the cause of the exit.

An alternative considered was to extend the Stopper.Stop() interface
somehow with a termination message, and continue to use the stopper as
the mechanism used by arbitrary modules to signal the desire to
terminate the server. I've decided against it, though, in favor of
introducing another mechanism for signaling requests to terminate up to
the cli layer, on the argument that the stopper is passed around so
widely and used in a lot of non-production libraries, such that it's
hard to trace and tell what production modules call Stop() on it and
under which circumstances. Also, stopper.Stop() cannot be deadlocks if
called from inside a stopper task (and our code generally runs inside
stopper tasks), so we'd also need to add a new async stop interface to
the stopper.
Thus, the goal of this patch is to keep the number of callers to
Stopper.Stop() small - right now there are two of them: one in the cli,
and one in the drainServer. In a followup patch, I might remove the
drainServer call, and then we'd be left with exactly one caller to
Stopper.Stop().

Fixes #85540

Release note: None
Epic: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/90208)
<!-- Reviewable:end -->

andreimatei commented:
First two commits are merging in #90108.

@aadityasondhi I said I'd push a commit to your #88837, but then I thought that a different PR altogether is better. Let's let Rafa take a look and then perhaps you can kindly take it over. I think perhaps some tests you're adding in #88837 can move here, I'm not sure.

andreimatei committed:
cli: use serverStatus more

Use this new communication channel to pass the server reference to the
shutdown goroutine. Before this patch, the server reference was passed
separately from the information about whether the server startup
finished, although the two pieces of information were dependent on one
another; it was not trivial to convince oneself that the shutdown
goroutine was not using a nil server reference.

Release note: None

andreimatei committed:
cli: tighten a channel type

Minor cleanup.

Release note: None

andreimatei committed:
cli: muck with stopper ownership

This is a minor refactor making the contract around who should call
stopper.Stop() more visible. The goroutines starting the server and
shutting down the servers coordinate between them through a
`serverStatus` struct, and there's a point where the resposibility for
stopping the stopper goes from the former to the latter. This patch
makes the stopper flow through the serverStatus, so that this transfer
of resposibility is more apparent.

Closes #90233

Release note: None

andreimatei committed:
server: change signaling between the drainServer and cli

This patch gets rid of the stopper.Stop() call in the drainServer,
leaving a single stopper.Stop() call - in the cli layer. Before this
patch, a drain RPC that wanted to terminate the process called
stopper.Stop(), which signaled the cli's shutdown routine. This patch
switches the signaling mechanism to use the stopTrigger, which is
already used by some terminal errors. The shutdown goroutine no longer
needs to listen to the stopper, as it becomes the only one that stops
the stopper.

Release note: None

knz reviewed (commented):
The overall notification structure here now is pretty good. I like it.

What I didn't like was what you did to the shutdown code. I don't understand why there's a need for a conditional. Maybe this needs to be clarified better.

There's a few typing style things to look at too, see inline comments.

Also, is my reading right that this code does not significantly change the sql session hearbeating/liveness logic? I remember you explaining that there was problems with it, is this moving to a different/subsequent PR?

> Reviewed 1 of 1 files at r4, 11 of 16 files at r6, 1 of 2 files at r8, 7 of 7 files at r9, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-:-NF9HF9n3jhyxCuoQ1Xc:b-2glcpa)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @andreimatei)

___
*[`pkg/cli/start.go` line 688 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9FuFr6jVbBq67KJAe:-NF9FuFr6jVbBq67KJAf:b-2gs2s2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L688)):*
> ```Go
> }
> 
> // serverStartupError wraps errors encoutered during server startup.
> ```

Why was this needed instead of a "simple" `errors.Wrap(err, "error during server startup")`?
___
*[`pkg/cli/start.go` line 739 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9G4n0407JCCmhKAb5:-NF9G4n0407JCCmhKAb6:b-brpppe) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L739)):*
> ```Go
> // then the serverStatus does not take ownership of the stopper; the caller is
> // responsible for calling stopper.Stop().
> func (s *serverStatus) setStarted(server serverShutdownInterface, stopper *stop.Stopper) bool {
> ```

If you intend this change to fix #90233, please state so in the PR description.
___
*[`pkg/cli/start.go` line 804 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9GDuc909EovCM6wpf:-NF9GDudBEMKnEWrpCBv:b-hz6g6p) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L804)):*
> ```Go
> 			// stopper.Stop(), so there's no need (and it's also too late) for us to
> 			// call startShutdownAsync().
> 			log.StartAlwaysFlush()
> ```

`log.StartAlwaysFlush` should be called unconditionally.
___
*[`pkg/cli/start.go` line 808 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9GJlPCvwZRO6JGYwu:-NF9GJlPCvwZRO6JGYwv:bwx16di) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L808)):*
> ```Go
> 			// There's no point in draining if the server didn't even fully start.
> 			drain := shutdownRequest.ServerStartupErr == nil
> 			startShutdownAsync(serverStatusMu, stopWithoutDrain, drain)
> ```

This can also be called unconditionally.
___
*[`pkg/server/stop_trigger.go` line 39 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9Gfe2C_Q25wJmGPxr:-NF9Gfe2C_Q25wJmGPxs:bypk23x) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/server/stop_trigger.go#L39)):*
> ```Go
> 	defer s.mu.Unlock()
> 	if s.mu.shutdownRequest != (ShutdownRequest{}) {
> 		// Someone else already triggered the shutdown.
> ```

I think this should log the additional request, we probably want to know if there were multiple shutdown requests for different reasons during troubleshooting.
___
*[`pkg/server/stop_trigger.go` line 48 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9GmR23blYgWeEpUzM:-NF9GmR23blYgWeEpUzN:bwt9ke2) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/server/stop_trigger.go#L48)):*
> ```Go
> }
> 
> type ShutdownRequest struct {
> ```

Pls doc this.
___
*[`pkg/server/stop_trigger.go` line 50 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9GvOJ3TYSwbg8fETZ:-NF9GvOKBXy7jUP7T7CA:bpbzj2k) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/server/stop_trigger.go#L50)):*
> ```Go
> type ShutdownRequest struct {
> 	DrainRPC         bool
> 	ServerStartupErr error
> ```

This feels like it would be better served with a 3-state enum + just 1 error field.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90208)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @knz)

___
*[`pkg/cli/start.go` line 688 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9FuFr6jVbBq67KJAe:-NF9vXxm-Qlc6DxwzGyW:bkpiy9q) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L688)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Why was this needed instead of a "simple" `errors.Wrap(err, "error during server startup")`?
</blockquote></details>

it was needed in the patch that introduced because the shutdown routine recognized the new error type, but the need goes away once we have the `ShutdownRequest` - so removed now.
___
*[`pkg/cli/start.go` line 739 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9G4n0407JCCmhKAb5:-NF9cnl8DqqjmTdFt60Q:b-bg62l5) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L739)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

If you intend this change to fix #90233, please state so in the PR description.
</blockquote></details>

It's in the commit message of the respective commit.
I'll leave the PR description be another commit message, if you don't mind.
___
*[`pkg/cli/start.go` line 804 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9GDuc909EovCM6wpf:-NF9fZoU0pj4N7V1eXGE:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L804)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

`log.StartAlwaysFlush` should be called unconditionally.
</blockquote></details>

done
___
*[`pkg/cli/start.go` line 808 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9GJlPCvwZRO6JGYwu:-NF9f_QVAlFkZ470UQ8F:bdcpjmg) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L808)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

This can also be called unconditionally.
</blockquote></details>

done, nice
___
*[`pkg/server/stop_trigger.go` line 39 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9Gfe2C_Q25wJmGPxr:-NF9vUYGBnukt1gPK1cE:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/server/stop_trigger.go#L39)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

I think this should log the additional request, we probably want to know if there were multiple shutdown requests for different reasons during troubleshooting.
</blockquote></details>

done
___
*[`pkg/server/stop_trigger.go` line 48 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9GmR23blYgWeEpUzM:-NF9ukKF8FEhsWOPHQi1:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/server/stop_trigger.go#L48)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

Pls doc this.
</blockquote></details>

done
___
*[`pkg/server/stop_trigger.go` line 50 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9GvOJ3TYSwbg8fETZ:-NF9ulBH3-k5xyH6ZIRg:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/server/stop_trigger.go#L50)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

This feels like it would be better served with a 3-state enum + just 1 error field.
</blockquote></details>

done


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> Also, is my reading right that this code does not significantly change the sql session hearbeating/liveness logic? I remember you explaining that there was problems with it, is this moving to a different/subsequent PR?

Yeah, there's a lot of problems with the liveness code. I'm working on it elsewhere, which is part of the reason I want to get this patch out of the way.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-:-NFAG4FH-MHmrfH6ckhz:b-uw2i09)* status: :shipit: complete! 0 of 0 LGTMs obtained (waiting on @aadityasondhi and @knz)



<!-- Sent from Reviewable.io -->

knz reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/>

> Reviewed 9 of 15 files at r16, 1 of 2 files at r18, 8 of 8 files at r19, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-:-NFDFud75gB3aYokJDzk:b593vm9)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi)

___
*[`pkg/cli/start.go` line 688 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9FuFr6jVbBq67KJAe:-NFDFfDB27dd9PgBYO2b:b-ft895p) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L688)):*
<details><summary><i>Previously, andreimatei (Andrei Matei) wrote…</i></summary><blockquote>

it was needed in the patch that introduced because the shutdown routine recognized the new error type, but the need goes away once we have the `ShutdownRequest` - so removed now.
</blockquote></details>

So I think you can remove this type now.


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90208)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi and @knz)

___
*[`pkg/cli/start.go` line 688 at r9](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-NF9FuFr6jVbBq67KJAe:-NFEUm1l5NhrDkp0s8_d:b-lhs69o) ([raw file](https://github.com/cockroachdb/cockroach/blob/cd6420d63494671483c91cd0be4b0f06c34a542c/pkg/cli/start.go#L688)):*
<details><summary><i>Previously, knz (Raphael 'kena' Poss) wrote…</i></summary><blockquote>

So I think you can remove this type now.
</blockquote></details>

done, thanks


<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
I think this is ready to merge, but Rafa please take a look at the changes to `testserver.go` in r26. I was previously missing hooking up the new shutdown signaling mechanism for `TestServers`, which was a problem for example for `\demo restart <x>`.
TFTR!

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-:-NFKFVOW0cBhz3ydQWZe:bsxdepm)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi, @herkolategan, @knz, and @renatolabs)



<!-- Sent from Reviewable.io -->

knz reviewed (commented):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> cheers

> Reviewed 1 of 16 files at r26, 4 of 9 files at r29.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-:-NFKOxWDCgG4GbixE7kZ:bvsg7at)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @aadityasondhi, @herkolategan, and @renatolabs)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
TFTR!

bors r_

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-:-NFLhXid1Hn5tIM4CiEj:b-oduyj1)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi, @herkolategan, @knz, and @renatolabs)



<!-- Sent from Reviewable.io -->

andreimatei reviewed (commented):
bors r+

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/90208#-:-NFLhaQP9kIckqzSZ0nx:b-ydi9ay)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @aadityasondhi, @herkolategan, @knz, and @renatolabs)



<!-- Sent from Reviewable.io -->
